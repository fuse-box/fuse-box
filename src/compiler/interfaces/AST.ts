import { IASTScope } from '../Visitor/Visitor';

export interface ASTNode {
  raw?: string;
  type: keyof typeof ASTType;

  $assign_pattern?: boolean;
  $fuse_classInitializers?: Array<ASTNode>;
  $fuse_class_declaration_visited?: boolean;
  $fuse_decoratorForClassIdentifier?: string;
  $fuse_decoratorsCaptured?: boolean;
  $fuse_decoratorsProcessed?: boolean;
  $fuse_visited?: boolean;
  $transpiled?: boolean;
  alternate?: ASTNode;
  base?: ASTNode;
  chain?: ASTNode;
  consequent?: ASTNode;
  decorators?: Array<ASTNode>;
  definite?: boolean;
  elementTypes?: Array<ASTNode>;
  elements?: Array<ASTNode>;
  expressions?: Array<ASTNode>;
  initializer?: ASTNode;
  literal?: ASTNode;
  moduleReference?: ASTNode;
  optional?: boolean;
  prefix?: boolean;
  returnType?: ASTNode;
  scope?: IASTScope;
  shorthand?: boolean;
  source?: ASTNode;
  sourceType?: string;
  specifiers?: Array<ASTNode>;
  test?: ASTNode;
  typeAnnotation?: ASTNode;
  typeName?: ASTNode;
  loc?: { end: { column: number; line: number }; start: { column: number; line: number } };

  imported?: ASTNode;
  members?: Array<ASTNode>;
  method?: boolean;

  JSXElement?: ASTNode;

  abstract?: boolean;
  accessibility?: string;
  argument?: ASTNode;
  arguments?: Array<ASTNode>;
  async?: boolean;
  attributes?: Array<ASTNode>;
  body?: ASTNode | Array<ASTNode>;
  callee?: ASTNode;
  children?: Array<ASTNode>;
  closingElement?: ASTNode;
  computed?: boolean;
  declaration?: ASTNode;
  declarations?: Array<ASTNode>;
  declare?: boolean;
  exported?: ASTNode;
  expression?: any;
  generator?: boolean;
  id?: ASTNode;
  implements?: Array<ASTNode>;
  init?: null | ASTNode;
  key?: ASTNode;
  kind?: string;
  left?: ASTNode;
  local?: ASTNode;
  name?: any;
  object?: ASTNode;
  openingElement?: ASTNode;
  operator?: string;
  parameter?: ASTNode;
  params?: Array<ASTNode>;
  properties?: Array<ASTNode>;
  property?: ASTNode;
  quasi?: ASTNode;
  quasis?: Array<ASTNode>;
  right?: ASTNode;
  static?: boolean;
  superClass?: ASTNode;
  tag?: ASTNode;
  tail?: ASTNode;
  value?: any;
}

export const ASTType = {
  EnumDeclaration: 'TSEnumDeclaration',
  EnumMember: 'TSEnumMember',

  ParameterProperty: 'TSParameterProperty',
  StringKeyword: 'TSStringKeyword',
  TypeAnnotation: 'TSTypeAnnotation',

  AbstractClassProperty: 'TSAbstractClassProperty',
  AbstractMethodDefinition: 'TSAbstractMethodDefinition',
  AsExpression: 'TSAsExpression',
  DeclareFunction: 'TSDeclareFunction',
  DeclareKeyword: 'TSDeclareKeyword',
  InterfaceDeclaration: 'TSInterfaceDeclaration',
  NonNullExpression: 'TSNonNullExpression',
  TypeAliasDeclaration: 'TSTypeAliasDeclaration',
  TypeAssertion: 'TSTypeAssertion',

  ArrayExpression: 'ArrayExpression',
  ArrayPattern: 'ArrayPattern',
  ArrowFunctionExpression: 'ArrowFunctionExpression',
  AssignmentExpression: 'AssignmentExpression',
  AssignmentPattern: 'AssignmentPattern',
  AwaitExpression: 'AwaitExpression',
  BigIntLiteral: 'BigIntLiteral',
  BinaryExpression: 'BinaryExpression',
  Block: 'Block',
  BlockStatement: 'BlockStatement',
  BreakStatement: 'BreakStatement',
  CallExpression: 'CallExpression',
  CatchClause: 'CatchClause',
  ClassBody: 'ClassBody',
  ClassDeclaration: 'ClassDeclaration',
  ClassExpression: 'ClassExpression',
  ClassProperty: 'ClassProperty',
  ConditionalExpression: 'ConditionalExpression',
  ContinueStatement: 'ContinueStatement',
  DebuggerStatement: 'DebuggerStatement',
  Decorator: 'Decorator',
  DoWhileStatement: 'DoWhileStatement',
  EmptyStatement: 'EmptyStatement',
  ExportAllDeclaration: 'ExportAllDeclaration',
  ExportDefaultDeclaration: 'ExportDefaultDeclaration',
  ExportNamedDeclaration: 'ExportNamedDeclaration',
  ExportSpecifier: 'ExportSpecifier',
  ExpressionStatement: 'ExpressionStatement',
  ForInStatement: 'ForInStatement',
  ForOfStatement: 'ForOfStatement',
  ForStatement: 'ForStatement',
  FunctionDeclaration: 'FunctionDeclaration',
  FunctionExpression: 'FunctionExpression',
  HTMLClose: 'HTMLClose',
  HTMLOpen: 'HTMLOpen',
  Identifier: 'Identifier',
  IfStatement: 'IfStatement',
  Import: 'Import',
  ImportDeclaration: 'ImportDeclaration',
  ImportDefaultSpecifier: 'ImportDefaultSpecifier',
  ImportNamespaceSpecifier: 'ImportNamespaceSpecifier',
  ImportSpecifier: 'ImportSpecifier',
  JSXAttribute: 'JSXAttribute',
  JSXClosingElement: 'JSXClosingElement',
  JSXClosingFragment: 'JSXClosingFragment',
  JSXElement: 'JSXElement',
  JSXEmptyExpression: 'JSXEmptyExpression',
  JSXExpressionContainer: 'JSXExpressionContainer',
  JSXFragment: 'JSXFragment',
  JSXIdentifier: 'JSXIdentifier',
  JSXMemberExpression: 'JSXMemberExpression',
  JSXOpeningElement: 'JSXOpeningElement',
  JSXOpeningFragment: 'JSXOpeningFragment',
  JSXSpreadAttribute: 'JSXSpreadAttribute',
  JSXSpreadChild: 'JSXSpreadChild',
  JSXText: 'JSXText',
  LabeledStatement: 'LabeledStatement',
  Line: 'Line',
  Literal: 'Literal',
  LogicalExpression: 'LogicalExpression',
  MemberExpression: 'MemberExpression',
  MetaProperty: 'MetaProperty',
  MethodDefinition: 'MethodDefinition',
  NewExpression: 'NewExpression',
  ObjectExpression: 'ObjectExpression',
  ObjectPattern: 'ObjectPattern',
  OptionalCallExpression: 'OptionalCallExpression',
  OptionalMemberExpression: 'OptionalMemberExpression',
  Program: 'Program',
  Property: 'Property',
  RestElement: 'RestElement',
  ReturnStatement: 'ReturnStatement',
  SequenceExpression: 'SequenceExpression',
  SpreadElement: 'SpreadElement',
  Super: 'Super',
  SwitchCase: 'SwitchCase',
  SwitchStatement: 'SwitchStatement',
  TaggedTemplateExpression: 'TaggedTemplateExpression',
  TemplateElement: 'TemplateElement',
  TemplateLiteral: 'TemplateLiteral',
  ThisExpression: 'ThisExpression',
  ThrowStatement: 'ThrowStatement',
  TryStatement: 'TryStatement',
  module: 'module',
  script: 'script',

  AbstractKeyword: 'AbstractKeyword',

  AnyKeyword: 'TSAnyKeyword',
  ArrayType: 'TSArrayType',

  AsyncKeyword: 'AsyncKeyword',
  BigIntKeyword: 'BigIntKeyword',
  BooleanKeyword: 'TSBooleanKeyword',
  CallSignatureDeclaration: 'CallSignatureDeclaration',
  ClassImplements: 'ClassImplements',
  ConditionalType: 'ConditionalType',
  ConstructSignatureDeclaration: 'ConstructSignatureDeclaration',
  ConstructorType: 'ConstructorType',
  EmptyBodyFunctionExpression: 'EmptyBodyFunctionExpression',
  ExportAssignment: 'ExportAssignment',

  ExportKeyword: 'ExportKeyword',
  ExternalModuleReference: 'ExternalModuleReference',
  FunctionType: 'TSFunctionType',
  ImportEqualsDeclaration: 'TSImportEqualsDeclaration',
  ImportExpression: 'ImportExpression',
  ImportType: 'ImportType',
  IndexSignature: 'IndexSignature',
  IndexedAccessType: 'IndexedAccessType',
  InferType: 'InferType',

  InterfaceBody: 'InterfaceBody',
  InterfaceHeritage: 'InterfaceHeritage',
  IntersectionType: 'IntersectionType',
  LiteralType: 'TSLiteralType',
  MappedType: 'MappedType',
  MethodSignature: 'MethodSignature',
  ModuleBlock: 'ModuleBlock',
  ModuleDeclaration: 'TSModuleDeclaration',
  NamespaceExportDeclaration: 'NamespaceExportDeclaration',
  NeverKeyword: 'TSNeverKeyword',

  NullKeyword: 'TSNullKeyword',
  NumberKeyword: 'TSNumberKeyword',
  ObjectKeyword: 'ObjectKeyword',
  OptionalType: 'OptionalType',

  ParenthesizedType: 'ParenthesizedType',
  PrivateKeyword: 'PrivateKeyword',
  PropertySignature: 'PropertySignature',
  ProtectedKeyword: 'ProtectedKeyword',
  PublicKeyword: 'PublicKeyword',
  QualifiedName: 'QualifiedName',
  ReadonlyKeyword: 'ReadonlyKeyword',
  RestType: 'RestType',
  StaticKeyword: 'StaticKeyword',

  SymbolKeyword: 'SymbolKeyword',
  ThisType: 'ThisType',
  TupleType: 'TSTupleType',

  JSDocNullableType: 'JSDocNullableType',
  JSDocUnknownType: 'JSDocUnknownType',
  TypeLiteral: 'TypeLiteral',
  TypeOperator: 'TypeOperator',
  TypeParameter: 'TypeParameter',
  TypeParameterDeclaration: 'TypeParameterDeclaration',
  TypeParameterInstantiation: 'TypeParameterInstantiation',
  TypePredicate: 'TypePredicate',
  TypeQuery: 'TypeQuery',
  TypeReference: 'TSTypeReference',
  UnaryExpression: 'UnaryExpression',
  UndefinedKeyword: 'TSUndefinedKeyword',
  UnionType: 'UnionType',
  UnknownKeyword: 'UnknownKeyword',
  UpdateExpression: 'UpdateExpression',
  VariableDeclaration: 'VariableDeclaration',
  VariableDeclarator: 'VariableDeclarator',
  VoidKeyword: 'TSVoidKeyword',
  WhileStatement: 'WhileStatement',
  WithStatement: 'WithStatement',
  YieldExpression: 'YieldExpression',
};
