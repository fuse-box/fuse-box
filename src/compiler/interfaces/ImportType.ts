export enum ImportType {
  REQUIRE,
  FROM,
  RAW_IMPORT,
  DYNAMIC,
}
