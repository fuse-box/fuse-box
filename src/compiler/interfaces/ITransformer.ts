import { ICompilerOptions } from '../../compilerOptions/interfaces';
import { Context } from '../../core/context';
import { IModule } from '../../moduleResolver/module';
import { IProductionContext } from '../../production/ProductionContext';
import { IVisit, IVisitorMod } from '../Visitor/Visitor';
import { ASTNode } from './AST';
import { ImportType } from './ImportType';

export type ITranformerCallback = (visit: IVisit) => undefined | IVisitorMod | void;

export interface ITransformerVisitors {
  onEachNode?: ITranformerCallback;
  onTopLevelTraverse?: ITranformerCallback;
}

export interface IRequireStatementModuleOptions {
  breakDependantsCache?: boolean;
}
export interface ITransformerCommon {
  compilerOptions: ICompilerOptions;
  ctx: Context;
  module: IModule;
  onRequireCallExpression?: (
    importType: ImportType,
    node: ASTNode,
    moduleOptions?: IRequireStatementModuleOptions,
  ) => void;
}

export interface ITransformerProduction {
  ctx: Context;
  module: IModule;
  productionContext: IProductionContext;
}

export interface ITransformer {
  target?: {
    test?: RegExp;
    type?: 'js' | 'js_ts' | 'ts';
  };
  commonVisitors?: (props?: ITransformerCommon) => ITransformerVisitors;
  productionWarmupPhase?: (props?: ITransformerProduction) => ITransformerVisitors;
}
