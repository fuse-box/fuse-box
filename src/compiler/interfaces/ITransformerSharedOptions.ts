import { ImportType } from './ImportType';
import { ASTNode } from './AST';

export interface ITransformerSharedOptions {
  onRequireCallExpression?: (importType: ImportType, node: ASTNode) => void;
}
