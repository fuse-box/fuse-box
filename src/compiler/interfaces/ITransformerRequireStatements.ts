import { ASTNode } from './AST';
import { IRequireStatementModuleOptions } from './ITransformer';
import { ImportType } from './ImportType';

export interface ITransformerRequireStatement {
  importType: ImportType;
  moduleOptions?: IRequireStatementModuleOptions;
  statement: ASTNode;
  value?: string;
}
export type ITransformerRequireStatementCollection = Array<ITransformerRequireStatement>;
