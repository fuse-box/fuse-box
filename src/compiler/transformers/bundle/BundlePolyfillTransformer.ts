import * as path from 'path';
import { ASTNode } from '../../interfaces/AST';
import { ITransformerSharedOptions } from '../../interfaces/ITransformerSharedOptions';
import { GlobalContext } from '../../program/GlobalContext';
import { createRequireStatement } from '../../Visitor/helpers';
import { IVisit, IVisitorMod } from '../../Visitor/Visitor';
import { ImportType } from '../../interfaces/ImportType';
import { ITarget } from '../../../config/PrivateConfig';

export interface IBundleEssentialProps {
  target?: ITarget;
  env?: { [key: string]: any };
  isBrowser?: boolean;
  isServer?: boolean;
  moduleFileName?: string;
}

export type IBundleEssntialTransformerOptions = IBundleEssentialProps & ITransformerSharedOptions;

export const PolyfillEssentialConfig = {
  stream: 'stream',
  buffer: 'buffer',
  Buffer: 'buffer',
  http: 'http',
  https: 'https',
};

export function BundlePolyfillTransformer(options: IBundleEssntialTransformerOptions) {
  const RequireStatementsInserted = {};
  const VariablesInserted = {};
  options = options || {};
  return (visit: IVisit): IVisitorMod => {
    const { node } = visit;

    if (visit.isLocalIdentifier) {
      const globalContext = visit.globalContext as GlobalContext;
      const locals = visit.scope && visit.scope.locals ? visit.scope.locals : {};
      const name = node.name;

      if (locals[name] || globalContext.hoisted[name]) {
        return;
      }

      switch (name) {
        case '__dirname':
          return { replaceWith: { type: 'Literal', value: options.moduleFileName } };
        case '__filename':
          return { replaceWith: { type: 'Literal', value: path.dirname(options.moduleFileName) } };
      }

      if (options.target !== 'browser') return;

      /**
       * *********************************************************************************
       * Polyfills stream buffer e.t.c
       */
      if (PolyfillEssentialConfig[name]) {
        if (VariablesInserted[name]) return;
        const statements: Array<ASTNode> = [];
        const moduleName = PolyfillEssentialConfig[name];
        if (!RequireStatementsInserted[moduleName]) {
          RequireStatementsInserted[moduleName] = 1;
          const statement = createRequireStatement(moduleName, name);
          if (options.onRequireCallExpression)
            options.onRequireCallExpression(ImportType.REQUIRE, statement.reqStatement);
          statements.push(statement.statement);
        }
        if (name !== moduleName) {
          statements.push({
            type: 'VariableDeclaration',
            kind: 'var',
            declarations: [
              {
                type: 'VariableDeclarator',
                init: {
                  type: 'Identifier',
                  name: moduleName,
                },
                id: {
                  type: 'Identifier',
                  name: name,
                },
              },
            ],
          });
        }
        if (statements.length) {
          return { prependToBody: statements };
        }
      }
    }
    return;
  };
}
