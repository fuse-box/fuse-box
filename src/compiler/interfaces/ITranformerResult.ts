import { ASTNode } from './AST';
import { ITransformerRequireStatementCollection } from './ITransformerRequireStatements';

export interface ITransformerResult {
  ast?: ASTNode;
  requireStatementCollection: ITransformerRequireStatementCollection;
}
