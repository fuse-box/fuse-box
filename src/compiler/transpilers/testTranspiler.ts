import * as buntis from 'buntis';
import { generate } from '../generator/generator';
import { ASTNode } from '../interfaces/AST';
import { ICompilerOptions } from '../interfaces/ICompilerOptions';
import { ImportType } from '../interfaces/ImportType';
import { ITransformerRequireStatementCollection } from '../interfaces/ITransformerRequireStatements';
import { BundlePolyfillTransformer, IBundleEssentialProps } from '../transformers/bundle/BundlePolyfillTransformer';
import { GlobalContextTransformer } from '../transformers/GlobalContextTransformer';
import { DynamicImportTransformer } from '../transformers/shared/DynamicImportTransformer';
import { ExportTransformer } from '../transformers/shared/ExportTransformer';
import { ImportTransformer } from '../transformers/shared/ImportTransformer';
import { JSXTransformer } from '../transformers/shared/JSXTransformer';
import { ClassConstructorPropertyTransformer } from '../transformers/ts/ClassConstructorPropertyTransformer';
import { CommonTSfeaturesTransformer } from '../transformers/ts/CommonTSfeaturesTransformer';
import { EnumTransformer } from '../transformers/ts/EnumTransformer';
import { NamespaceTransformer } from '../transformers/ts/NameSpaceTransformer';
import { IVisit, IVisitorMod } from '../Visitor/Visitor';

import { BrowserProcessTransformer } from '../transformers/bundle/BrowserProcessTransformer';
import { ITransformerList, transpileModule } from '../program/transpileModule';
import { createGlobalContext } from '../program/GlobalContext';
import { RequireStatementInterceptor } from '../transformers/bundle/RequireStatementInterceptor';
import { BundleFastConditionUnwrapper } from '../transformers/bundle/BundleFastConditionTransformer';

export interface ICompileModuleProps {
  code: string;
  globalContext?: any;
  transformers?: Array<(globalContext) => (visit: IVisit) => IVisitorMod>;
  compilerOptions?: ICompilerOptions;
  bundleProps?: IBundleEssentialProps;
}

export function testTranspile(props: ICompileModuleProps) {
  const ast = buntis.parseTSModule(props.code, {
    directives: true,
    jsx: true,
    next: true,
    loc: true,
  });
  const requireStatementCollection: ITransformerRequireStatementCollection = [];
  function onRequireCallExpression(importType: ImportType, statement: ASTNode) {
    // making sure we have haven't emitted the same property twice
    if (!statement['emitted']) {
      Object.defineProperty(statement, 'emitted', { enumerable: false, value: true });
      requireStatementCollection.push({ importType, statement });
    }
  }

  let bundleProps: IBundleEssentialProps = props.bundleProps || {
    moduleFileName: './somefile.ts',
    target: 'browser',
  };

  const defaultTransformers: ITransformerList = [
    GlobalContextTransformer(),
    BundleFastConditionUnwrapper({
      env: bundleProps.env,
      isBrowser: bundleProps.isBrowser,
      isServer: bundleProps.isServer,
    }),
    RequireStatementInterceptor({ onRequireCallExpression }),
    bundleProps.target === 'browser' && BrowserProcessTransformer({ ...bundleProps, onRequireCallExpression }),
    BundlePolyfillTransformer({ ...bundleProps, onRequireCallExpression }),
    BundlePolyfillTransformer({ onRequireCallExpression, ...bundleProps }),
    DynamicImportTransformer({ onRequireCallExpression }),
    EnumTransformer(),
    ClassConstructorPropertyTransformer(),
    JSXTransformer(),
    NamespaceTransformer(),

    // must be before export/import
    CommonTSfeaturesTransformer(),
    ImportTransformer({ onRequireCallExpression }),
    ExportTransformer({ onRequireCallExpression }),
  ];
  transpileModule({
    ast: ast as ASTNode,
    compilerOptions: props.compilerOptions,
    globalContext: createGlobalContext(props.globalContext),
    transformers: defaultTransformers,
  });
  //console.log(JSON.stringify(ast, null, 2));
  const res = generate(ast, {});

  return { code: res, requireStatementCollection };
}
