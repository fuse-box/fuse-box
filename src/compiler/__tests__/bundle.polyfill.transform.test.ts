import { ImportType } from '../interfaces/ImportType';
import { PolyfillEssentialConfig } from '../transformers/bundle/BundlePolyfillTransformer';
import { testTranspile } from '../transpilers/testTranspiler';

describe('Bundle polyfill transform test', () => {
  describe('Common transform ', () => {
    it('should replace __dirname', () => {
      const result = testTranspile({
        code: `
           console.log(__dirname)
      `,
        bundleProps: {
          moduleFileName: '/some-dir/file.ts',
        },
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should replace __filename', () => {
      const result = testTranspile({
        code: `
           console.log(__filename)
      `,
        bundleProps: {
          moduleFileName: '/some-dir/file.ts',
        },
      });
      expect(result.code).toMatchSnapshot();
    });
  });

  describe('polyfills', () => {
    describe('Single reference', () => {
      for (const name in PolyfillEssentialConfig) {
        const moduleName = PolyfillEssentialConfig[name];
        it(`shuold insert ${moduleName}`, () => {
          const result = testTranspile({
            code: `console.log(${moduleName})`,
          });
          expect(result.requireStatementCollection).toEqual([
            {
              importType: ImportType.REQUIRE,
              statement: {
                type: 'CallExpression',
                callee: { type: 'Identifier', name: 'require' },
                arguments: [{ type: 'Literal', value: moduleName }],
              },
            },
          ]);
          expect(result.code).toMatchSnapshot();
        });
      }
    });

    describe('Member reference reference', () => {
      for (const name in PolyfillEssentialConfig) {
        const moduleName = PolyfillEssentialConfig[name];
        it(`shuold insert ${moduleName} with method reference`, () => {
          const result = testTranspile({
            code: `console.log(${moduleName}.method)`,
          });

          expect(result.code).toMatchSnapshot();
        });
      }
    });
    describe('Member reference with a call', () => {
      for (const name in PolyfillEssentialConfig) {
        const moduleName = PolyfillEssentialConfig[name];
        it(`shuold insert ${moduleName} with method reference with a call`, () => {
          const result = testTranspile({
            code: `console.log(${moduleName}.method() )`,
          });

          expect(result.code).toMatchSnapshot();
        });
      }
    });

    describe('Should not add anything', () => {
      it('should not add Buffer because its been hoisted', () => {
        const result = testTranspile({
          code: `
            exports.Buffer = Buffer;
            function Buffer(){}
          `,
        });

        expect(result.code).toMatchSnapshot();
      });

      it('should not add Buffer because its has been defined', () => {
        const result = testTranspile({
          code: `
            function some(){
              const buffer = {};
              console.log(buffer)
            }
          `,
        });
        expect(result.code).toMatchSnapshot();
      });

      it('should not add Buffer because it has NOT been defined', () => {
        const result = testTranspile({
          code: `
            function some(){
              console.log(buffer)
            }
          `,
        });

        expect(result.code).toMatchSnapshot();
      });
    });
  });
});
