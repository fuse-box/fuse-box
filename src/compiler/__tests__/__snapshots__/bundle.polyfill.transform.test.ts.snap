// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bundle polyfill transform test Common transform  should replace __dirname 1`] = `
"console.log(\\"/some-dir/file.ts\\");
"
`;

exports[`Bundle polyfill transform test Common transform  should replace __filename 1`] = `
"console.log(\\"/some-dir\\");
"
`;

exports[`Bundle polyfill transform test polyfills Member reference reference shuold insert buffer with method reference 1`] = `
"var buffer = require(\\"buffer\\");
console.log(buffer.method);
"
`;

exports[`Bundle polyfill transform test polyfills Member reference reference shuold insert buffer with method reference 2`] = `
"var buffer = require(\\"buffer\\");
console.log(buffer.method);
"
`;

exports[`Bundle polyfill transform test polyfills Member reference reference shuold insert http with method reference 1`] = `
"var http = require(\\"http\\");
console.log(http.method);
"
`;

exports[`Bundle polyfill transform test polyfills Member reference reference shuold insert https with method reference 1`] = `
"var https = require(\\"https\\");
console.log(https.method);
"
`;

exports[`Bundle polyfill transform test polyfills Member reference reference shuold insert stream with method reference 1`] = `
"var stream = require(\\"stream\\");
console.log(stream.method);
"
`;

exports[`Bundle polyfill transform test polyfills Member reference with a call shuold insert buffer with method reference with a call 1`] = `
"var buffer = require(\\"buffer\\");
console.log(buffer.method());
"
`;

exports[`Bundle polyfill transform test polyfills Member reference with a call shuold insert buffer with method reference with a call 2`] = `
"var buffer = require(\\"buffer\\");
console.log(buffer.method());
"
`;

exports[`Bundle polyfill transform test polyfills Member reference with a call shuold insert http with method reference with a call 1`] = `
"var http = require(\\"http\\");
console.log(http.method());
"
`;

exports[`Bundle polyfill transform test polyfills Member reference with a call shuold insert https with method reference with a call 1`] = `
"var https = require(\\"https\\");
console.log(https.method());
"
`;

exports[`Bundle polyfill transform test polyfills Member reference with a call shuold insert stream with method reference with a call 1`] = `
"var stream = require(\\"stream\\");
console.log(stream.method());
"
`;

exports[`Bundle polyfill transform test polyfills Should not add anything should not add Buffer because it has NOT been defined 1`] = `
"var buffer = require(\\"buffer\\");
function some() {
  console.log(buffer);
}
"
`;

exports[`Bundle polyfill transform test polyfills Should not add anything should not add Buffer because its been hoisted 1`] = `
"exports.Buffer = Buffer;
function Buffer() {}
"
`;

exports[`Bundle polyfill transform test polyfills Should not add anything should not add Buffer because its has been defined 1`] = `
"function some() {
  const buffer = {};
  console.log(buffer);
}
"
`;

exports[`Bundle polyfill transform test polyfills Single reference shuold insert buffer 1`] = `
"var buffer = require(\\"buffer\\");
console.log(buffer);
"
`;

exports[`Bundle polyfill transform test polyfills Single reference shuold insert buffer 2`] = `
"var buffer = require(\\"buffer\\");
console.log(buffer);
"
`;

exports[`Bundle polyfill transform test polyfills Single reference shuold insert http 1`] = `
"var http = require(\\"http\\");
console.log(http);
"
`;

exports[`Bundle polyfill transform test polyfills Single reference shuold insert https 1`] = `
"var https = require(\\"https\\");
console.log(https);
"
`;

exports[`Bundle polyfill transform test polyfills Single reference shuold insert stream 1`] = `
"var stream = require(\\"stream\\");
console.log(stream);
"
`;
